vmui: enhancement of autocomplete feature

Loori-R opened the pull request:
In this pull request, we have extended the autocomplete functionality by adding hints and suggestions for:

1. **MetricSQL Functions** 
   - A comprehensive list of `MetricSQL` functions along with their descriptions has been added.
   - Note: Autocompletion for function arguments is not supported at this time.

2. **Labels** 
   - Autocomplete suggestions have been provided for `labels` associated with specific metrics.

3. **Values** 
   - Enhanced support for `value` suggestions in autocomplete.

Resolved issue #4993
See also issue #3006

| ![demo](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/29711459/410020ad-34dc-4d2d-8941-8d5eb71eaa18) | ![demo](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/29711459/eb3d8af7-92fa-4c18-9e09-b84479c75bae) |
|---------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------|
| ![demo](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/29711459/6ba2f5ef-c296-4f42-b70f-0b4781239f7f) | ![demo](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/29711459/151f64b8-c571-4e17-bf0a-26baa24f6cd0) |

hagen1778 reviewed (approved):
LGTM! 
But I noticed that sum functions in autocomplete have old descriptions.
For example, [rollup_rate](https://docs.victoriametrics.com/MetricsQL.html#rollup_rate) function seems outdated. How do we maintain those descriptions?

Loori-R commented:
@hagen1778 I took the description of the functions from the [documentation](https://docs.victoriametrics.com/MetricsQL.html#rollup_rate). I don't see any differences, maybe the documentation has an old description. Now the description is stored in a separate file on vmui. 
I will think about how to synchronize the description with [MetricsQL.md](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/docs/MetricsQL.md) to always keep it up to date.

For example:
<img width="1119" alt="Screenshot 2023-09-26 at 13 21 11" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/29711459/fe5dd32d-40ae-4d17-8ac0-4ffe645170d9">
<img width="520" alt="Screenshot 2023-09-26 at 13 21 24" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/29711459/053e94eb-e096-497e-b761-91dc6bbd1eaf">

hagen1778 commented:
Oh, I meant the following part:
> See [this article](https://valyala.medium.com/why-irate-from-prometheus-doesnt-capture-spikes-45f9896d7832) in order to understand better when to use rollup_rate().
Optional 2nd argument "min", "max" or "avg" can be passed to keep only one calculation result and without adding a label.
The calculations are performed individually per each time series returned from the given [series_selector](https://docs.victoriametrics.com/keyConcepts.html#filtering).
Metric names are stripped from the resulting rollups. Add [keep_metric_names](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/docs/MetricsQL.md#keep_metric_names) modifier in order to keep metric names.

Loori-R commented:
Ah, okay, I will update the PR to fetch the full description directly from MetricsQL.md.

Loori-R commented:
@hagen1778 updated

hagen1778 reviewed (commented):
I'm trying to build it and see following errors:

```
> vmui@0.1.0 update-metricsql
> ln -s ../../../../../../docs/MetricsQL.md src/assets/MetricsQL.md || true

ln: src/assets/MetricsQL.md: File exists
```

This ^was solved by adding `-f` flag to `ln` cmd in order to override destinations.

However, the next error isn't that easy:
```
> vmui@0.1.0 build
> GENERATE_SOURCEMAP=false react-app-rewired build

Creating an optimized production build...
Failed to compile.

Module not found: Error: Can't resolve '../assets/MetricsQL.md' in '/build/packages/vmui/src/hooks'
```

Loori-R commented:
Sorry, I didn't notice. I've corrected it. I replaced `ln` with `cp` because there was an issue with the file link.

dmitryk-dk commented on app/vmui/packages/vmui/src/components/Configurators/QueryEditor/QueryEditorAutocomplete.tsx:
```diff
@@ -0,0 +1,127 @@
+
+  const context = useMemo(() => {
+    [metricRegexp, labelRegexp, valueRegexp].forEach(regexp => regexp.lastIndex = 0);
+    if (valueRegexp.test(value)) {
+      return ContextType.value;
+    } else if (labelRegexp.test(value)) {
+      return ContextType.label;
+    } else if (metricRegexp.test(value)) {
+      return ContextType.metricsql;
+    }
+    return ContextType.empty;
```
nit: we can use switch as in code below

Loori-R replied:
updated

dmitryk-dk reviewed (approved):
LGTM! Added small comment

hagen1778 commented:
> Sorry, I didn't notice. I've corrected it. I replaced ln with cp because there was an issue with the file link.

Hey! Still doesn't work for me:
```

> vmui@0.1.0 update-metricsql
> cp ../../../../docs/MetricsQL.md src/assets/MetricsQL.md || true

cp: can't stat '../../../../docs/MetricsQL.md': No such file or directory

```

Maybe would be easier to copy file via Makefile?

Loori-R commented:
Added the copy command to the Makefile.

hagen1778 reviewed (approved):
Looks great!

valyala commented:
While the auto-complete feature looks great overall, I have the following concerns:

1. Make sure that vmui doesn't pre-load any information related to auto-completion when the autocomplete checkbox is disabled, since loading this information can significantly increase load on both VictoriaMetrics server and the VMUI itself.
2. It looks like VMUI pre-loads the list of available metric names via `/api/v1/label/__name__/values` endpoint on the initial page load when user didn't enter anything into the `query` field. This can significantly increase load on VictoriaMetrics server if it contains large number of unique metric names (for example, if it contains millions of unique Graphite metrics). This also can significantly slow down VMUI if it needs to pre-load millions of unique metric names. It would be great to limit the number of pre-loaded metric names, label names and label values in order to prevent from high resource usage at VMUI. VictoriaMetrics accepts an optional `limit` query arg at `/api/v1/labels` and `/api/v1/label/<labelName>/values` endpoints - see [these docs](https://docs.victoriametrics.com/#prometheus-querying-api-enhancements) for details. It would be great if VMUI would set the limit to some sensible value like 1000 or 10000, and then automatically disable the auto-completion if this limit is reached in the response from VictoriaMetrics.
3. The `/api/v1/labels` and `/api/v1/label/.../values` APIs may become very slow when the database contains hundreds of millions of unique time series because of [high churn rate](https://docs.victoriametrics.com/FAQ.html#what-is-high-churn-rate) and when `match[]` query arg is passed to these APIs. This may make the auto-complete feature unusable while simultaneously significantly increasing the load on VictoriaMetrics server.
4. It is better to specify `start` and `end` query args when querying `/api/v1/labels` and `/api/v1/label/.../values`, since by default VictoriaMetrics returns values from these APIs for time series seen during the last day. See https://docs.victoriametrics.com/#prometheus-querying-api-enhancements for details.

I personally don't like the current implementation because of the issues described above. An additional issue is very low signal/noise for the information provided by the auto-complete feature, e.g. it show a ton of information to the user, and this information blinks on every keystroke and generates non-trivial load on both VictoriaMetrics server and the user's web browser, while in reality the user needs only a small fraction of the displayed auto-complete information. We need to think better on how to reduce the load on VictoriaMetrics server and web browser, while simultaneously increasing the amounts of useful information displayed to the user. For example, how about requiring the user to press some hotkey (such as `cmd+space`) in order to trigger auto-complete instead if triggering it with every keystroke.

Loori-R commented:
@valyala 
1. If autocomplete is disabled, no queries related to it are loaded.
2. Added a `limit of 1000` for all queries related to autocomplete in PR #5253.
3. These queries are only executed if there is a full match with the metric name from a previously loaded list. For example, to execute the query `/api/v1/labels?match[]={metric}`, a metric from the list obtained from `/api/v1/label/__name__/values` must be entered in the field. Similarly, for `/api/v1/label/{label}/values?match[]={metric}`, the label must be in the list from `/api/v1/labels?match[]={metric}`. Therefore, the exact name of the metric/label needed to find the values should be in the `match[]`. Let me know if there is a more optimal solution.
4. Added start and end for all queries related to autocomplete in PR #5253.
5. Also added an on/off toggle via hotkey `cmd + shift + a`, as `cmd + space` is used for Spotlight on MacOS.


Regarding the "noise," I am not quite sure what you want autocomplete to display. If it's bothersome, it can be turned off. Currently, functions, metrics, labels, and values are displayed depending on the context. To reduce the load, a list of popular queries can be created and only those displayed, although I am not sure this would be more convenient.
